### PR TITLE
Add logit ref; allow non-refs to be called in refs; support torch function mode on torch.ops

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78014
* #78016
* #77854
* __->__ #77816

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -1014,7 +1003,7 @@ def native_batch_norm(
         mean = running_mean
         invstd = 1 / (torch.sqrt(running_var + eps))
         # Very annoying inconsistency where CPU and CUDA give different shapes
-        if input.device.type == "cuda":
+        if input.device.type != "cpu":
```
Comment (mruberry) [inline torch/_decomp/decompositions.py:1006]: Nice fix

File: torch/_prims/context.py [hunk 2]
```diff
@@ -156,3 +162,42 @@ def __torch_function__(
                 return fn(*args, **kwargs)  # type: ignore[operator]
 
         return func(*args, **kwargs)
+
+
+@functools.lru_cache
+def torch_to_refs_map():
```
Comment (mruberry) [inline torch/_prims/context.py:168]: Comment here

File: torch/_prims/context.py [hunk 3]
```diff
@@ -156,3 +162,42 @@ def __torch_function__(
...
+    r = {}
+    for mod_torch, mod_refs in modules:
+        for s in mod_refs.__all__:
+            r[mod_torch.__dict__.get(s)] = mod_refs.__dict__.get(s)
+    return r
+
+class TorchRefsMode(torch.overrides.TorchFunctionMode):
+    """
+    Switches the interpretation of torch.* functions and Tensor methods to
+    use PrimTorch refs in torch._refs.
```
Comment (mruberry) [inline torch/_prims/context.py:184]: One more sentence with a simple example would be nice, also note that direct calls to refs are preserved?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -18783,6 +18783,16 @@ def __init__(
...
         torch_opinfo_name="special.i1e",
     ),
+    ElementwiseUnaryPythonRefInfo(
+        "_refs.special.logit",
+        torch_opinfo_name="logit",
+        decorators=(
+            DecorateInfo(toleranceOverride({
+                torch.bfloat16: tol(atol=1e-2, rtol=0),
+            }), 'TestCommon', 'test_python_reference_consistency', device_type='cpu'),
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_reference_meta_functions',),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:18793]: Comment for this failure

**Review by mruberry — APPROVED** [review]
Cool! See a few comment suggestions inline

**Comment by ezyang** [comment]
While debugging the context manager I also noticed that modes didn't work with torch.ops calls, now they do.

**Pull request closed without merging**